pyomo/repn/linear.py

```diff
@@ -718,17 +680,26 @@ def _before_named_expression(visitor, child):
 
 
 def _before_expr_if(visitor, child):
+    # We want a _before_expr_if to catch constant test expressions so
+    # that we ONLY evaluate the appropriate branch.
     test, t, f = child.args
     if is_fixed(test):
         try:
-            test = test()
+            test = visitor._eval_expr(test)
         except:
             return True, None
-        subexpr = LinearRepnVisitor(
+        if test.__class__ is InvalidNumber:
+            return True, None
+        subexpr = visitor.__class__(
             visitor.subexpression_cache, visitor.var_map, visitor.var_order
         ).walk_expression(t if test else f)
         if subexpr.nonlinear is not None:
             return False, (_GENERAL, subexpr)
+        # This test is not needed for LINEAR, but is for the derived
+        # QUADRATIC.  As this is the ONLY _before_* specialization
```

Comment:
Long-term question: is there potential for other "before" specializations? If so, maybe it makes more sense to move this.